File: list_6/C/mastermind/main.c
//
// Created by serhi on 12/22/2025.
//

#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <stdbool.h>

#include "master_mind.h"

int main() {
    int *allPosibilitiesList = allPosibilities();

    bool *validMask = malloc(TOTALPERMUTATIONS * sizeof(bool));
    for (int i = 0; i < TOTALPERMUTATIONS; i++) validMask[i] = true;

    int attempts = 1;
    int rightPlace, wrongPlace;

    printf("Think of a sequence of %d numbers (1-%d).\n", GUESSNUMBERS, GUESSRANGE);
    printf("------------------------------------------------------------------\n");

    while (true) {
        int guessIndex = -1;
        for (int i = 0; i < TOTALPERMUTATIONS; i++) {
            if (validMask[i]) {
                guessIndex = i;
                break; // Found a valid guess
            }
        }

        if (guessIndex == -1) {
            printf("You are lying!\n");
            break;
        }

        int *currentGuess = &allPosibilitiesList[guessIndex * GUESSNUMBERS];

        printf("Attempt %d: ", attempts);
        printGuess(currentGuess);
        printf("\n");

        printf("On right place: ");
        scanf("%d", &rightPlace);
        printf("On wrong place: ");
        scanf("%d", &wrongPlace);


        if (rightPlace == GUESSNUMBERS) {
            printf("I guessed it in %d attempts!\n", attempts);
            break;
        }

        filterList(allPosibilitiesList, validMask, currentGuess, rightPlace, wrongPlace);

        int remaining = 0;
        for (int i = 0; i < TOTALPERMUTATIONS; i++) if (validMask[i]) remaining++;
        printf("Possible codes remaining: %d\n\n", remaining);

        attempts++;
    }

    free(allPosibilitiesList);
    free(validMask);

    return 0;
}

File: list_6/Python/mastermind/main.py
import master_mind

def printGuess(guess):
    print("[ ", end="")
    for i in range(master_mind.GUESSNUMBERS):
        print(f"{guess[i]} ", end="")
    print("] ", end="")

def main():
    allPosibilitiesList = master_mind.allPosibilities()

    validMask = [True] * master_mind.TOTALPERMUTATIONS

    attempts = 1

    print(f"Think of a sequence of {master_mind.GUESSNUMBERS} numbers (1-{master_mind.GUESSRANGE}).")
    print("I will try to guess it.")
    print("-" * 66)

    while True:
        guessIndex = -1
        for i in range(master_mind.TOTALPERMUTATIONS):
            if validMask[i]:
                guessIndex = i
                break

        if guessIndex == -1:
            print("Error: No valid possibilities left. You might have entered incorrect feedback previously.")
            break

        currentGuess = allPosibilitiesList[guessIndex]

        print(f"Attempt {attempts}: I guess ", end="")
        printGuess(currentGuess)
        print()

        print("On right place: ", end="")
        rightPlace = int(input())
        print("On wrong place: ", end="")
        wrongPlace = int(input())

        if rightPlace == master_mind.GUESSNUMBERS:
            print(f"I guessed it in {attempts} attempts!")
            break

        master_mind.filterList(allPosibilitiesList, validMask, currentGuess, rightPlace, wrongPlace)

        remaining = 0
        for i in range(master_mind.TOTALPERMUTATIONS):
            if validMask[i]:
                remaining += 1
        print(f"Possible codes remaining: {remaining}\n")

        attempts += 1

if __name__ == "__main__":
    main()